Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/NetworkSearchController.java

```diff
@@ -131,53 +146,104 @@ public String getSearch(@ProjectIdentifier @PathVariable("projectUnixName") Stri
     @InjectProjectByName
-    @RequestMapping(value = "sites/{projectUnixName}/networks/search", method = RequestMethod.GET)
-    public String getSearchTransformedNetwork(@ProjectIdentifier @PathVariable("projectUnixName") String projectUnixName,
+    @RequestMapping(value = "sites/{projectUnixName}/networks/search", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
+    public String submitSearchTransformedNetworkRequest(
+            @ProjectIdentifier @PathVariable("projectUnixName") String projectUnixName,
             @RequestParam("conceptId") String conceptId, @CheckAccess @InjectProject IProject project, Model model)
-                    throws QuadrigaStorageException {
-        
+            throws QuadrigaStorageException {
+
         String lemma = "";
         String searchNodeLabel = "";
-        
         // Fetch ConceptPower entries related to the conceptId
         IConcept concept = cpCache.getConceptByUri(conceptId);
-   
+
         if (concept != null) {
             searchNodeLabel = concept.getWord();
             lemma = concept.getDescription();
-           
         }
-     
-        ITransformedNetwork transformedNetwork = transformationManager
-                .getSearchTransformedNetwork(project.getProjectId(), conceptId, INetworkStatus.APPROVED);
 
-        String json = null;
-        if (transformedNetwork != null) {
-            json = jsonCreator.getJson(transformedNetwork.getNodes(), transformedNetwork.getLinks());
-        }
+        Callable<PublicSearchObject> callable = () -> {
 
-        if (transformedNetwork == null || transformedNetwork.getNodes().size() == 0) {
-            model.addAttribute("isNetworkEmpty", true);
+            PublicSearchObject publicSearchObject = new PublicSearchObject();
+            ITransformedNetwork transformedNetwork = transformationManager
+                    .getSearchTransformedNetwork(project.getProjectId(), conceptId, INetworkStatus.APPROVED);
+            if (transformedNetwork != null) {
+                publicSearchObject.setNodes(jsonCreator.getNodes(transformedNetwork.getNodes()));
+                publicSearchObject.setLinks(jsonCreator.getLinks(transformedNetwork.getLinks()));
+            }
+            if (transformedNetwork == null || transformedNetwork.getNodes().size() == 0) {
+                publicSearchObject.setNetworkEmpty(true);
+            }
+            return publicSearchObject;
+        };
+
+        Future<PublicSearchObject> future = executorService.submit(callable);
+        Random randomTokenGenerator = new Random();
+        Integer randomToken = randomTokenGenerator.nextInt(100);
+        while (searchResultMap.containsKey(randomToken)) {
+            randomToken = randomTokenGenerator.nextInt(100);
         }
-
-        model.addAttribute("jsonstring", json);
-        model.addAttribute("networkid", "\"\"");
-        model.addAttribute("project", project);
+        searchResultMap.put(randomToken, future);
```

Comment:
there should be a service that keeps track of running searches. This is not the job of the controller.

Reply:
also, your from the token that you use to check on the progress the original search string should be retrievable. you don't want to start the same search again, if a user searches for the same query again shortly after it was submitted.